lib/vnet/local_app_provider.go

```diff
@@ -0,0 +1,216 @@
+		return nil, errNoMatch
+	}
+
+	if isDescendantSubdomain(fqdn, profileName) {
+		// The queried app fqdn is a subdomain of this cluster proxy address.
+		return rootClient, nil
+	}
+
+	leafClusters, err := getLeafClusters(ctx, rootClient)
+	if err != nil {
+		// Good chance we're here because the user is not logged in to the profile.
+		log.ErrorContext(ctx, "Failed to list leaf clusters, apps in this cluster will not be resolved.", "profile", profileName, "error", err)
+		return nil, errNoMatch
+	}
+
+	// Prefix with an empty string to represent the root cluster.
+	allClusters := append([]string{""}, leafClusters...)
+	for _, leafClusterName := range allClusters {
+		clusterClient, err := p.ClientApplication.GetCachedClient(ctx, profileName, leafClusterName)
+		if err != nil {
+			log.ErrorContext(ctx, "Failed to get cluster client, apps in this cluster will not be resolved.", "profile", profileName, "leaf_cluster", leafClusterName, "error", err)
+			continue
+		}
+
+		clusterConfig, err := p.clusterConfigCache.GetClusterConfig(ctx, clusterClient)
+		if err != nil {
+			log.ErrorContext(ctx, "Failed to get VnetConfig, apps in the cluster will not be resolved.", "profile", profileName, "leaf_cluster", leafClusterName, "error", err)
+			continue
+		}
+		for _, zone := range clusterConfig.DNSZones {
+			if isDescendantSubdomain(fqdn, zone) {
+				return clusterClient, nil
+			}
+		}
+	}
+	return nil, errNoMatch
+}
+
+var errNoMatch = errors.New("cluster does not match queried FQDN")
+
+func getLeafClusters(ctx context.Context, rootClient ClusterClient) ([]string, error) {
+	var leafClusters []string
+	nextPage := ""
+	for {
+		remoteClusters, nextPage, err := rootClient.CurrentCluster().ListRemoteClusters(ctx, 0, nextPage)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		for _, rc := range remoteClusters {
+			leafClusters = append(leafClusters, rc.GetName())
+		}
+		if nextPage == "" {
+			return leafClusters, nil
+		}
+	}
+}
+
+func (p *localAppProvider) resolveAppInfoForCluster(
+	ctx context.Context,
+	clusterClient ClusterClient,
+	profileName, leafClusterName, fqdn string,
+) (*vnetv1.AppInfo, error) {
+	log := log.With("profile", profileName, "leaf_cluster", leafClusterName, "fqdn", fqdn)
+	// An app public_addr could technically be full-qualified or not, match either way.
+	expr := fmt.Sprintf(`(resource.spec.public_addr == "%s" || resource.spec.public_addr == "%s") && hasPrefix(resource.spec.uri, "tcp://")`,
+		strings.TrimSuffix(fqdn, "."), fqdn)
+	resp, err := apiclient.GetResourcePage[types.AppServer](ctx, clusterClient.CurrentCluster(), &proto.ListResourcesRequest{
+		ResourceType:        types.KindAppServer,
+		PredicateExpression: expr,
+		Limit:               1,
+	})
+	if err != nil {
+		// Don't return an unexpected error so we can try to find the app in different clusters or forward the
+		// request upstream.
+		log.InfoContext(ctx, "Failed to list application servers", "error", err)
+		return nil, errNoTCPHandler
```

Comment:
Since on error this method always logs `err` and then returns `errNoTCPHandler`, could we instead move this behavior up, always return plain `err` from `resolveAppInfoForCluster` and document why we're not using `trace.Wrap`?

Reply:
thanks for calling it out because on taking another look, after we know there's a matching app, i think this function should actually be returning the errors. updated to do so